tpm2/kdf.go

```diff
@@ -28,43 +25,66 @@ import (
 // See: https://trustedcomputinggroup.org/resource/tpm-library-specification/
 // The key & label parameters must not be zero length, but contextU &
 // contextV may be.
-// Only SHA1 & SHA256 hash algorithms are implemented at this time.
+// The label parameter is a non-null-terminated string.
 func KDFa(hashAlg Algorithm, key []byte, label string, contextU, contextV []byte, bits int) ([]byte, error) {
-	var counter uint32
-	remaining := (bits + 7) / 8 // As per note at the bottom of page 44.
-	var out []byte
-
-	var mac hash.Hash
-	switch hashAlg {
-	case AlgSHA1:
-		mac = hmac.New(sha1.New, key)
-	case AlgSHA256:
-		mac = hmac.New(sha256.New, key)
-	default:
-		return nil, fmt.Errorf("hash algorithm 0x%x is not supported", hashAlg)
+	h, err := hashAlg.HashConstructor()
+	if err != nil {
+		return nil, err
 	}
+	mac := hmac.New(h, key)
 
-	for remaining > 0 {
-		counter++
-		if err := binary.Write(mac, binary.BigEndian, counter); err != nil {
-			return nil, fmt.Errorf("pack counter: %v", err)
-		}
+	out := kdf(mac, bits, func() {
 		mac.Write([]byte(label))
 		mac.Write([]byte{0}) // Terminating null character for C-string.
 		mac.Write(contextU)
 		mac.Write(contextV)
-		if err := binary.Write(mac, binary.BigEndian, uint32(bits)); err != nil {
-			return nil, fmt.Errorf("pack bits: %v", err)
-		}
-
-		out = mac.Sum(out)
-		remaining -= mac.Size()
-		mac.Reset()
-	}
+		binary.Write(mac, binary.BigEndian, uint32(bits))
+	})
+	return out, nil
+}
 
-	if len(out) > bits/8 {
-		out = out[:bits/8]
+// KDFe implements TPM 2.0's ECDH key derivation function, as defined in
+// section 11.4.9.3 of the TPM revision 2 specification part 1.
+// See: https://trustedcomputinggroup.org/resource/tpm-library-specification/
+// The z parameter is the x coordinate of one parties private ECC key and the other parties public ECC key.
+// The partyUInfo and partyVInfo are the x coordinates of the initiators and the responders ECC points respectively.
+// The use parameter is a non-null-terminated string.
```

Comment:
Here and above, be sure to document parameters in their function order: `z`, `use`, etc...